libia2/ia2.c

```diff
@@ -21,65 +41,81 @@ int protect_pages(struct dl_phdr_info *info, size_t size, void *data) {
     Elf64_Addr start = info->dlpi_addr + phdr.p_vaddr;
     Elf64_Addr end = start + phdr.p_memsz;
     if (start <= address && address <= end) {
-      this_compartment = true;
-      break;
+      return true;
     }
   }
-  if (!this_compartment) {
+  return false;
+}
+
+/// Map ELF segment flags to mprotect access flags
+static int segment_flags_to_access_flags(Elf64_Word flags) {
+  return ((flags & PF_X) != 0 ? PROT_EXEC : 0) |
+         ((flags & PF_W) != 0 ? PROT_WRITE : 0) |
+         ((flags & PF_R) != 0 ? PROT_READ : 0);
```

Comment:
You dropped `| PROT_NONE`

Reply:
PROT_NONE is 0, we don't need to OR that in here.